consensus/istanbul/backend/announce.go

```diff
@@ -589,130 +649,311 @@ func (sb *Backend) regossipAnnounce(msg *istanbul.Message, payload []byte) error
+				Address: entry.Address,
+				Version: entry.Version,
+				Signature: entry.Signature,
+			})
+			sb.lastSignedAnnounceVersionsGossiped[entry.Address] = time.Now()
 		}
-		sb.cachedAnnounceMsgsMu.RUnlock()
 	}
+	sb.lastSignedAnnounceVersionsGossipedMu.Unlock()
+	if len(signedAnnVersionsToRegossip) > 0 {
+		return sb.gossipSignedAnnounceVersionsMsg(signedAnnVersionsToRegossip)
+	}
+	return nil
+}
+
+func (sb *Backend) updateAnnounceVersion() {
+	sb.updateAnnounceVersionCh <- struct{}{}
+	<-sb.updateAnnounceVersionCompleteCh
+}
 
-	if len(announcesToRequest) > 0 {
-		announcesToRequestBytes, err := rlp.EncodeToBytes(announcesToRequest)
+// setAndShareUpdatedAnnounceVersion generates a new enode certificate message and sends
+// it to this node's proxy (if applicable).
+// TODO: When a generated announce no longer creates a new version, the version
+// parameter can be removed and instead have the version generated in this function.
+// Tracked here: https://github.com/celo-org/celo-monorepo/issues/2668
+func (sb *Backend) setAndShareUpdatedAnnounceVersion(version uint) error {
+	logger := sb.logger.New("func", "updateAnnounceVersion")
+	// Send new versioned enode msg to all other registered or elected validators
+	validatorConnSet, err := sb.retrieveValidatorConnSet()
+	if err != nil {
+		return err
+	}
+	// Don't send any messages if this node is not registered or elected
+	if !validatorConnSet[sb.Address()] {
+		logger.Trace("Not registered or elected, not updating announce version")
+		return nil
+	}
+	destAddresses := make([]common.Address, len(validatorConnSet))
+	i := 0
+	for address := range validatorConnSet {
+		destAddresses[i] = address
+		i++
+	}
+	enodeCertificateMsg, err := sb.generateEnodeCertificateMsg(version)
+	if err != nil {
+		return err
+	}
+	sb.setEnodeCertificateMsg(enodeCertificateMsg)
+	// Send the new versioned enode msg to the proxy peer
+	if sb.config.Proxied && sb.proxyNode != nil && sb.proxyNode.peer != nil {
+		err := sb.sendEnodeCertificateMsg(sb.proxyNode.peer, enodeCertificateMsg)
 		if err != nil {
-			logger.Error("Error encoding announce to request array", "announcesToRequest", common.ConvertToStringSlice(announcesToRequest), "err", err)
+			logger.Error("Error in sending versioned enode msg to proxy", "err", err)
 			return err
 		}
-
-		logger.Trace("Going to send a GetAnnounces", "announcesToRequest", common.ConvertToStringSlice(announcesToRequest), "peer", peer)
-
-		go peer.Send(istanbulGetAnnouncesMsg, announcesToRequestBytes)
+	}
+	payload, err := enodeCertificateMsg.Payload()
+	if err != nil {
+		return err
+	}
+	err = sb.Multicast(destAddresses, payload, istanbulEnodeCertificateMsg)
+	if err != nil {
+		return err
 	}
 
-	return nil
+	// Generate and gossip a new signed announce version
+	newSignedAnnVersion, err := sb.generateSignedAnnounceVersion(version)
+	if err != nil {
+		return err
+	}
+	return sb.upsertSignedAnnounceVersionEntries([]*vet.SignedAnnounceVersionEntry{
```

Comment:
`sb.upsertSignedAnnounceVersionEntries` has quite a few side effects and it is unclear if those are intended here.

Reply:
Just renamed it to `upsertAndGossipSignedAnnounceVersionEntries` and moved `startGossipAnnounceTask` out of there, hope that makes it clearer.